test/StringUtils.test.cpp

```diff
@@ -116,60 +129,83 @@ TEST(String, stringFrom)
 	EXPECT_EQ("129"s, NAS2D::stringFrom<unsigned char>(129));
 	EXPECT_EQ("-1"s, NAS2D::stringFrom<signed char>(-1));
 
-	EXPECT_EQ("0"s, NAS2D::stringFrom<unsigned short>(0));
-	EXPECT_EQ("1"s, NAS2D::stringFrom<unsigned short>(1));
-
-	EXPECT_EQ("0"s, NAS2D::stringFrom<unsigned int>(0));
-	EXPECT_EQ("1"s, NAS2D::stringFrom<unsigned int>(1));
-
-	EXPECT_EQ("0"s, NAS2D::stringFrom<unsigned long>(0));
-	EXPECT_EQ("1"s, NAS2D::stringFrom<unsigned long>(1));
-
-	EXPECT_EQ("0"s, NAS2D::stringFrom<unsigned long long>(0));
-	EXPECT_EQ("1"s, NAS2D::stringFrom<unsigned long long>(1));
-
-	EXPECT_EQ("-1"s, NAS2D::stringFrom<short>(-1));
-	EXPECT_EQ("0"s, NAS2D::stringFrom<short>(0));
-	EXPECT_EQ("1"s, NAS2D::stringFrom<short>(1));
-
-	EXPECT_EQ("-1"s, NAS2D::stringFrom<int>(-1));
-	EXPECT_EQ("0"s, NAS2D::stringFrom<int>(0));
-	EXPECT_EQ("1"s, NAS2D::stringFrom<int>(1));
-
-	EXPECT_EQ("-1"s, NAS2D::stringFrom<long>(-1));
-	EXPECT_EQ("0"s, NAS2D::stringFrom<long>(0));
-	EXPECT_EQ("1"s, NAS2D::stringFrom<long>(1));
-
-	EXPECT_EQ("-1"s, NAS2D::stringFrom<long long>(-1));
-	EXPECT_EQ("0"s, NAS2D::stringFrom<long long>(0));
-	EXPECT_EQ("1"s, NAS2D::stringFrom<long long>(1));
-
-	EXPECT_EQ("-1.000000"s, NAS2D::stringFrom<float>(-1));
-	EXPECT_EQ("-1.000000"s, NAS2D::stringFrom<float>(-1.0));
-	EXPECT_EQ("-1.100000"s, NAS2D::stringFrom<float>(-1.1f));
-	EXPECT_EQ("0.000000"s, NAS2D::stringFrom<float>(0));
-	EXPECT_EQ("0.000000"s, NAS2D::stringFrom<float>(0.0));
-	EXPECT_EQ("1.000000"s, NAS2D::stringFrom<float>(1));
-	EXPECT_EQ("1.000000"s, NAS2D::stringFrom<float>(1.0));
-	EXPECT_EQ("1.100000"s, NAS2D::stringFrom<float>(1.1f));
-
-	EXPECT_EQ("-1.000000"s, NAS2D::stringFrom<double>(-1));
-	EXPECT_EQ("-1.000000"s, NAS2D::stringFrom<double>(-1.0));
-	EXPECT_EQ("-1.100000"s, NAS2D::stringFrom<double>(-1.1));
-	EXPECT_EQ("0.000000"s, NAS2D::stringFrom<double>(0));
-	EXPECT_EQ("0.000000"s, NAS2D::stringFrom<double>(0.0));
-	EXPECT_EQ("1.000000"s, NAS2D::stringFrom<double>(1));
-	EXPECT_EQ("1.000000"s, NAS2D::stringFrom<double>(1.0));
-	EXPECT_EQ("1.100000"s, NAS2D::stringFrom<double>(1.1));
-
-	EXPECT_EQ("-1.000000"s, NAS2D::stringFrom<long double>(-1));
-	EXPECT_EQ("-1.000000"s, NAS2D::stringFrom<long double>(-1.0));
-	EXPECT_EQ("-1.100000"s, NAS2D::stringFrom<long double>(-1.1L));
-	EXPECT_EQ("0.000000"s, NAS2D::stringFrom<long double>(0));
-	EXPECT_EQ("0.000000"s, NAS2D::stringFrom<long double>(0.0));
-	EXPECT_EQ("1.000000"s, NAS2D::stringFrom<long double>(1));
-	EXPECT_EQ("1.000000"s, NAS2D::stringFrom<long double>(1.0));
-	EXPECT_EQ("1.100000"s, NAS2D::stringFrom<long double>(1.1L));
+}
+
+
+template<typename T>
+class StringFromSignedIntegerTest : public ::testing::Test
+{
+  public:
+	  StringFromSignedIntegerTest<T>()
+	{}
+
+	T negOne{-1};
+	T zero{0};
+	T one{1};
+};
+
+using StringFromSignedIntegerTestTypes = ::testing::Types<short, int, long, long long>;
+TYPED_TEST_SUITE(StringFromSignedIntegerTest, StringFromSignedIntegerTestTypes);
```

Comment:
It might be possible to use parenthesis, if you want to avoid the `using` alias:
```cpp
TYPED_TEST_SUITE(StringFromSignedIntegerTest, (::testing::Types<short, int, long, long long>));
```

Reply:
Documentation says it's required to make the macro work because of the commas.

Reply:
I realize that. I don't think the documentation is entirely accurate.

Without the parenthesis, it will certain give an error due to the commas.

Reply:
It  errors for the following change:

```
//using StringFromSignedIntegerTestTypes = ::testing::Types<short, int, long, long long>;
TYPED_TEST_SUITE(StringFromSignedIntegerTest, (::testing::Types<short, int, long, long long>));
```

Intellisense:
`Function definition 'TYPED_TEST_SUITE' not found. expected a type specifier` (remember, if it can be parsed as a function, it will. Macros are just text replacement, after all.)

Compiler freaks out:
```
1>E:\Git\nas2d-core\test\StringUtils.test.cpp(140,1): error C2059: syntax error: '>'
1>E:\Git\nas2d-core\test\StringUtils.test.cpp(140,1): error C2062: type 'unknown-type' unexpected
1>E:\Git\nas2d-core\test\StringUtils.test.cpp(140,1): error C2039: 'type': is not a member of '`global namespace''
1>E:\Git\nas2d-core\test\StringUtils.test.cpp(142,1): error C2065: 'gtest_type_params_StringFromSignedIntegerTest_': undeclared identifier
1>E:\Git\nas2d-core\test\StringUtils.test.cpp(142,1): error C2923: 'testing::internal::TypeParameterizedTest': 'gtest_type_params_StringFromSignedIntegerTest_' is not a valid template type argument for parameter 'Types'
1>E:\Git\nas2d-core\test\StringUtils.test.cpp(142,1): error C2955: 'testing::internal::TypeParameterizedTest': use of class template requires template argument list
1>E:\Git\vcpkg\installed\x86-windows\include\gtest\internal\gtest-internal.h(689): message : see declaration of 'testing::internal::TypeParameterizedTest'
1>E:\Git\nas2d-core\test\StringUtils.test.cpp(142,1): error C2672: 'testing::internal::GenerateNames': no matching overloaded function found
1>E:\Git\nas2d-core\test\StringUtils.test.cpp(142,1): error C2974: 'testing::internal::GenerateNames': invalid template argument for 'Types', type expected
1>E:\Git\vcpkg\installed\x86-windows\include\gtest\internal\gtest-internal.h(675): message : see declaration of 'testing::internal::GenerateNames'
1>E:\Git\vcpkg\installed\x86-windows\include\gtest\internal\gtest-internal.h(698,1): fatal error C1903: unable to recover from previous error(s); stopping compilation
1>INTERNAL COMPILER ERROR in 'C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.25.28610\bin\HostX86\x86\CL.exe'
1>    Please choose the Technical Support command on the Visual C++
1>    Help menu, or open the Technical Support help file for more information
1>Done building project "test.vcxproj" -- FAILED.
```